README.md
# GANce - `gance` 

![Sample network output](./docs_assets/gance_sample.png)

Maps music and video into the latent space of StyleGAN (mostly focused on StyleGAN2) networks.

See [esologic.com/gance](https://www.esologic.com/gance) for the development story, 
hardware configuration, future plans etc.

## Demo

The following is a snippet from the music video for 'buzzz' by 
[Won Pound](https://wonpound.bandcamp.com/) from his self-titled album _Won Pound_ released by 
[Minaret Records](https://www.minaretrecords.com/):

https://user-images.githubusercontent.com/3516293/176499620-8605cc84-02e9-4906-96f9-6c20f9faa795.mp4

Because of GitHub's video size limitations, this clip has been compressed. The complete, 
album-length music video for _Won Pound_ is available in 4k on YouTube 
[here](https://youtu.be/qc573jxXgII). The 'buzzz' section of the video starts 
[here](https://youtu.be/qc573jxXgII?t=1079).

The video was entirely synthesized with GANce, here's the command to produce
[NOVA](https://youtu.be/1ced1KxJRl4), for example:

```bash
python music_into_networks.py projection-file-blend \
  --wav './gance/assets/audio/masters/15 - NOVA (end 2) (pass 2).wav' \
  --output-path ./nova.mp4 \
  --debug-path ./nova_debug.mp4 \
  --debug-side-length 1000 \
  --debug-window 200 \
  --run-config ./nova_config.json \
  --log log.txt \
  --projection-file-path ./gance/assets/projection_files/resumed_prod_nova_3-1.hdf5 \
  --networks-json ./prod_networks.json \
  --output-fps 60 \
  --output-side-length 2160 \
  --alpha 0.25 \
  --fft-roll-enabled \
  --fft-amplitude-range -5 5 \
  --blend-depth 12 \
  --phash-distance 25 \
  --bbox-distance 50 \
  --track-length 5
```

## Usage

The functionality available in the libraries are exposed through a series of command line 
interfaces. The following is a listing of the commands that are stable and mostly usable at this
point.

| **Script**                 | **Description**                                                                                                                                                                                                                                          |
|----------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `check_move_networks.py`   | Scan directories of networks for pickled networks, for each of these networks, load them, and feed in a vector to make sure they are still functional (no bit rot).                                                                                      |
| `music_into_networks.py`   | Feed inputs (music, videos) into a network and record the output. Also tools to visualize these vectors against the network outputs.                                                                                                                     |
| `process_images.py`        | Tools to prepare images for styleGAN training.                                                                                                                                                                                                           |
| `project_video_to_file.py` | Tools to [project](https://github.com/NVlabs/stylegan2#projecting-images-to-latent-space) video into the latent space of a given network, and also visualize the results. Resulting projection files are consumed primarily in `music_into_networks.py`. |
| `snythesize_images.py`     | Given some styleGAN networks, load each network and synthesize a number of images with them. Interesting vectors can be reused with other networks.                                                                                                      |

## Getting Started

### Python Dependencies

See the `requirements` directory for required Python modules for building, testing, developing etc.
They can all be installed in a [virtual environment](https://docs.python.org/3/library/venv.html) 
using the follow commands:

```
python3 -m venv venv
source venv/bin/activate
pip install -r ./requirements/dev.txt -r ./requirements/prod.txt -r ./requirements/test.txt
```

There's also a bin script to do this:

```
./tools/create_venv.sh
```


## Developer Guide

The following is documentation for developers that would like to contribute
to GANce.

### Type System Limitations

A common idiom found in this application is the following:

```python
# Shape (1, Any, Any)
CompleteLatentsType = NewType("CompleteLatentsType", "np.ndarray[np.float32]")  # type: ignore
```

Where a `NewType` is created based on an `np.ndarray`. This doesn't really do anything outside
of making the programs a bit more simple to understand to the reader. The following would pass
`mypy`:

```python
CompleteLatentsType("oh no!")
```

This is because:
* We're stuck on `numpy==1.16.4` because of the `tensorflow` dependency of `stylegan2`.
* `np.ndarray` types are `Any` as far as `mypy` is concerned in this version.
* When you `NewType("MyType", Any)`, it stops `mypy` from being able to identify type mismatches.

Read more here: https://github.com/python/mypy/issues/6701

We could resolve this if we take the time to verify that it's okay to upgrade the `numpy` version.


### Testing

This project uses pytest to manage and run unit tests. Unit tests located in the `test` directory 
are automatically run during the CI build. You can run them manually with:

```
./tools/run_tests.sh
```

### Local Linting

There are a few linters/code checks included with this project to speed up the development process:

* Black - An automatic code formatter, never think about python style again.
* Isort - Automatically organizes imports in your modules.
* Pylint - Check your code against many of the python style guide rules.
* Mypy - Check your code to make sure it is properly typed.

You can run these tools automatically in check mode, meaning you will get an error if any of them
would not pass with:

```
./tools/run_checks.sh
```

Or actually automatically apply the fixes with:

```
./tools/apply_linters.sh
```

There are also scripts in `./tools/` that include run/check for each individual tool.


### Using pre-commit

First you need to init the repo as a git repo with:

```
git init
```

Then you can set up the git hook scripts with:

```
pre-commit install
```

By default:

* black
* pylint
* isort
* mypy

Are all run in apply-mode and must pass in order to actually make the commit.

Also by default, pytest needs to pass before you can push.

If you'd like skip these checks you can commit with:

```
git commit --no-verify
```

If you'd like to quickly run these pre-commit checks on all files (not just the staged ones) you
can run:

```
pre-commit run --all-files
```



CHANGELOG.md
# Changelog

0.22.0 - (2022-06-29)
------------------

* Added new param `--debug-side-length` to `music_into_network.py` commands to control the vertical
size of the debug videos.
* Switched canonical image resizing interpolation to `cv2.INTER_CUBIC`, resulting in better looking
output videos.
* Note: This version that produced the production _Won Pound_ music videos.


0.21.0 - (2022-04-17)
------------------

* Fixed `--log` argument in `music_into_network.py` commands.
* Fixed ability to set output side length in `music_into_network.py` commands.
* Added alternate frames -> video writer functionality that invokes ffmpeg directly, for better
control of output quality. "High Quality" settings are optimized for uploading big videos to 
YouTube. Default settings still use MP4V encoder.

  
0.20.0 - (2022-03-29)
------------------

* In `music_into_models.py projection-file-blend`, the notion of connecting overlay appearances
with audio activity has been removed, meaning `--complexity-change-rolling-sum-window` and 
`--complexity-change-threshold` are disabled. This functionality was underdeveloped, and didn't 
really work in a predictable way. The functionality will still be there in 
`projection_file_blend_api`, just not accessible via the CLI.
* In the same script, if `--phash-distance`, `--bbox-distance` and `--track-length` aren't provided, 
overlays won't be computed in the resulting videos.


0.19.0 - (2022-03-06)
------------------

* Removed underscores from all CLI arguments.
* Refactored codebase to make sure 'network' was used over 'model' in all cases. They're networks
not models.
A few more changes to projection file blending:
* Added a few more arguments to `music_into_models.py` to be able to make selecting the production
list of models easier. Now input args can be saved to file, and lists of models can be passed in
as json.
* Switched phash computation to only take place on the bounding box regions of the images. 
This results to tighter looking overlays.
* Can use `--wav` argument multiple times now, good for working through projection files with
multiple tracks.
* Audio files are now scaled in duration to the length of the video contained in the projection file.
This way, the number of FFTs, and the number of final latents to blend are always the same. There's
a check to make sure the projected frame count and the # of vectors are not too far off. 


0.18.0 - (2022-02-08)
------------------

* For music/video synthesis, introduced concept of the overlay, which is taking regions of the
target images from a projection file, and displaying them over the synthesized output from the
network. This gives the viewer a clue as to how the video was generated, and creates another
interesting dimension of the resulting videos.
* Refactored synthesis pipeline invoked in `music_into_models.py` as a series of iterators. This
enabled introducing the overlay concept while still being able to efficiently use memory. This
stream processing approach required a few modifications to existing concepts to be able to handle
streams rather than complete lists etc.
* Introduced video writers that can write video/audio to a file, and then forward the resulting 
stream of frames to be re-used. 
* Broke out some more standard functionality for dealing with images and videos. Introduced
a set of types for single frames, and streams of frames (videos) as well as resolutions etc.
* The `visualize_final_latents` command in `read_projection_file.py` can now accept multiple input 
audio files that will be appended to each other and then added to the resulting video.
* Finally, resolved the dlib import problem by creating `faces.py`, which enables the import of the
offending library right before use, rather than on loading.


0.17.0 - (2021-11-22)
------------------

* Breaking up project and renaming it `GANce` for publication on GitHub.
* Added a number of examples for a blog post explaining this 
project [here](https://www.esologic.com/introducing-gance/).


0.16.1 - (2021-10-28)
------------------

* Fixed a bug in synthesis file reader/writer. Both now use the standard vector types. 


0.16.0 - (2021-10-23)
------------------

* This version will create the final projection files for the yellow album cover project.
* In `project_video_to_file.py`, added ability to process multiple input videos at once. Either
a directory of videos or multiple videos passed via CLI. Added unit testing for this as well.
* Created a few projection specific docker containers in `docker-compose.yml` based on the `develop`
model.
* Added a very small cli, `read_projection_file.py` to read a projection file and turn the final
images vs. target images into a video.


0.15.0 - (2021-10-17)
------------------

* In `project_video_to_file.py`, added ability to override the fps of the input video. Use with
caution.
* Added type system for key data types. Images, Vectors/Matrices, Latents etc. See comment in
readme but these are mostly to aid in documentation.
* Cleaned up repo structure a bit during refactor to add types, but still some things that need to
be deleted.
* Added integration style testing of projection file reader and writer that require GPU.

0.14.0 - (2021-10-12)
------------------

* Added visualization/numeric analysis functionality to decide the projection steps and projection
FPS that will be used in the production projections.
* Added a number of helper functions to make working with projection files more expedient.
* Fixed bug in `synthesis_file_into_models.py`


0.13.0 - (2021-09-20)
------------------

* Added a CLI to, given a directory of models, create a number of random output images, some with
and some without faces. This tool, `images_from_models.py` will be used to create the prod album
cover.
* Introduced notion of a synthesis file, which is a json file produced alongside the images 
previously mentioned which records the associated vector and model used to create the image.

batch_runner.py
# pylint: skip-file

import os
from pathlib import Path
from typing import List, NamedTuple

from jinja2 import Template


class BatchVariation(NamedTuple):
    wavs: List[str]
    projection_file_path: str
    name: str


if __name__ == "__main__":

    inputs = [
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/15 - NOVA (end 2) (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/resumed_prod_nova_3-1.hdf5",
            name="nova",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/01 - yellow intro (pass 2).wav",
                "./gance/assets/audio/masters/02 - sub domain grass 4th (pass 2).wav",
                "./gance/assets/audio/masters/03 - NOD3 .work (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_intro_sub_domain_node_work-2.hdf5",
            name="intro_subdomain_nodework",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/04 - downinthe_(56r FLK of) (pass 2).wav",
                "./gance/assets/audio/masters/05 - familiar___+ (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_down_in_the_familiar_cropped-1.hdf5",
            name="down_in_the_familiar",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/06 - leaveit - pass 02 (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/resumed_prod_leave_it-1.hdf5",
            name="leave_it",
        ),
        BatchVariation(
            wavs=["./gance/assets/audio/masters/07 - fourth dance theme (pass 2).wav"],
            projection_file_path="./gance/assets/projection_files/third_fourth.hdf5",
            name="fourth_dance_theme",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/08 - foam pt. 2 (pass 2).wav",
                "./gance/assets/audio/masters/09 - Forethought xTIT(an) (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_foam_part_2_forthought_titan-1.hdf5",
            name="foam_forethought_titan",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/10 - balloons (pop) (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_ballons-1.hdf5",
            name="ballons",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/11 - town (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/2_resumed_prod_town_1.hdf5",
            name="town",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/12 - buzzz (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/resumed_buzzzz_1.hdf5",
            name="buzzz",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/13 - change - pass 02 (pass 2).wav",
                "./gance/assets/audio/masters/14 - usb hard disk - pass 02 (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_change_usb_hard_disk-1.hdf5",
            name="change_usb_hard_disk",
        ),
        BatchVariation(
            wavs=[
                "./gance/assets/audio/masters/16 - birds of solaris (pass 2).wav",
            ],
            projection_file_path="./gance/assets/projection_files/prod_birds_of_solaris-1.hdf5",
            name="birds_of_solaris",
        ),
    ]

    output_dir = Path("./gance/assets/output/final_runs/final_final_max_quality/")
    output_dir.mkdir(exist_ok=True)

    commands = [
        " ".join(
            [
                "./venv/bin/python /home/gpu/gance/music_into_networks.py projection-file-blend",
                *[f'--wav "{wav}"' for wav in command_parameters.wavs],
                f"--output-path {str(output_dir.joinpath(f'{command_parameters.name}.mp4'))}",
                f"--debug-path {str(output_dir.joinpath(f'{command_parameters.name}_debug.mp4'))}",
                f"--debug-window 200",
                f"--run-config {str(output_dir.joinpath(f'{command_parameters.name}.json'))}",
                f"--log log.txt",
                f"--projection-file-path {command_parameters.projection_file_path}",
                f"--networks-json ./prod_networks.json",
                f"--output-fps 60",
                f"--output-side-length 2160",
                f"--alpha 0.25",
                f"--fft-roll-enabled",
                f"--fft-amplitude-range -5 5",
                f"--blend-depth 12",
                f"--phash-distance 25",
                f"--bbox-distance 50",
                f"--track-length 5",
            ]
        )
        + os.linesep
        for command_parameters in inputs
    ]

    with open("./batch.sh.jinja2") as f:
        template = Template(f.read())

        with open("./batch_final.sh", "w") as f:
            f.write(template.render(commands=" ".join(commands)))
